mat: format complex dense matrices

wamuir opened the pull request:
Please take a look.

This PR adds a `CFormat` function in gonum/mat to format complex dense matrices; this function is intended to complement the current `Format` function.  Relevant issue #641 and a relevant PR #1414, although that PR appears to have stalled.  Further, PR #1414 takes the approach of wholesale duplicating `format.go`, which may not be preferred.

Summary:

Following https://github.com/gonum/gonum/pull/1414#issuecomment-666282217, an objective was to avoid code duplication while implementing a formatter for complex matrices.  To summarize the proposed approach:  value formatters are added for formatting floats and complex numbers, which are then used during matrix formatting.  Since these value formatters satisfy the `fmt.Formatter` interface, `fmt.State` can be passed, offering simplification in several areas (e.g., re-construction of a format string is no longer necessary).  As it would be inefficient to spin up a a new formatter for each element in a matrix, a value formatter holds a buffer for reuse.

To tackle widthing, a `state` type is proposed (satisfying `fmt.State`) that can be fit to a matrix and that stores information on column widths, using the existing widther type. As before, iterating through matrix elements is required.   After a call to `At`, a `state` will respond to subsequent `Width` calls with the relevant column width.  Recommendations toward improving implementation are appreciated.

Notwithstanding the need for additional internal types, this feels simpler to me on the whole and I believe this approach should make extending to new syntaxes or data types easier, if desired.  Thanks for reviewing -- will address feedback/comments. 

Bench:

I've included a few benchmarks below.  These compare the current and proposed implementation when formatting 10x10, 100x100 and 1000x1000 dense matrices.

| implementation | benchmark           |  iter |            time |          bytes |           alloc |
| :------------- | :------------------ | ----: | --------------: | -------------: | --------------: |
| current        | BenchmarkFormat10   | 36134 |     33063 ns/op |      2693 B/op |    16 allocs/op |
| current        | BenchmarkFormat100  |   358 |   3747403 ns/op |   1672025 B/op |   137 allocs/op |
| current        | BenchmarkFormat1000 |     3 | 334034817 ns/op | 163557800 B/op |  1066 allocs/op |
| proposed       | BenchmarkFormat10   | 22086 |     54554 ns/op |      2886 B/op |    19 allocs/op |
| proposed       | BenchmarkFormat100  |   170 |   7072273 ns/op |   1670752 B/op |    49 allocs/op |
| proposed       | BenchmarkFormat1000 |     2 | 697675398 ns/op | 163542896 B/op |    80 allocs/op |

sbinet commented:
just a drive-by comment:
could you run the benchmarks before/after the proposed optimization a couple of times (say 10-20) and show the result of `benchstat`?
with the limited statistics you presented in the commit message, I get this rather poor statistical power result:

```
$> benchstat ./ref.txt ./new.txt 
name        old time/op    new time/op    delta
Format10      33.1µs ± 0%    54.6µs ± 0%   ~     (p=1.000 n=1+1)
Format100     3.75ms ± 0%    7.07ms ± 0%   ~     (p=1.000 n=1+1)
Format1000     334ms ± 0%     698ms ± 0%   ~     (p=1.000 n=1+1)

name        old alloc/op   new alloc/op   delta
Format10      2.69kB ± 0%    2.89kB ± 0%   ~     (p=1.000 n=1+1)
Format100     1.67MB ± 0%    1.67MB ± 0%   ~     (p=1.000 n=1+1)
Format1000     164MB ± 0%     164MB ± 0%   ~     (p=1.000 n=1+1)

name        old allocs/op  new allocs/op  delta
Format10        16.0 ± 0%      19.0 ± 0%   ~     (p=1.000 n=1+1)
Format100        137 ± 0%        49 ± 0%   ~     (p=1.000 n=1+1)
Format1000     1.07k ± 0%     0.08k ± 0%   ~     (p=1.000 n=1+1)
```

ie:
```
$> go test -run=NONE -bench=Format -benchmem -count=20 |& tee ref.txt
## apply optimizations
$> go test -run=NONE -bench=Format -benchmem -count=20 |& tee new.txt
$> benchstat ./ref.txt ./new.txt
```

thanks.

PS: `benchstat` is available here:
- [golang.org/x/perf/benchstat](https://golang.org/x/perf/benchstat)

kortschak commented:
@wamuir Thank you for doing this. I've only taken a brief look, but it looks like nice work so far. I hope to be able to have a deeper look over the weekend.

kortschak commented on mat/format_test.go:
```diff
@@ -238,3 +240,16 @@ func TestFormat(t *testing.T) {
 	}
 }
+
+func benchmarkFormat(b *testing.B, size int) {
+	src := rand.NewSource(1)
+	a, _ := randDense(size, 1.0, src)
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		_ = fmt.Sprintf("%v", Formatted(a))
+	}
+}
+
+func BenchmarkFormat10(b *testing.B)   { benchmarkFormat(b, 10) }
+func BenchmarkFormat100(b *testing.B)  { benchmarkFormat(b, 100) }
+func BenchmarkFormat1000(b *testing.B) { benchmarkFormat(b, 1000) }
```
I'd be inclined to use sub-benchmarks for this. There are many places where we don't, but that is for historical reasons (the didn't exist when those benchmarks were written).

kortschak commented on mat/format.go:
```diff
@@ -19,7 +19,7 @@ func Formatted(m Matrix, options ...FormatOption) fmt.Formatter {
 		o(&f)
 	}
-	return f
+	return &f
```
This change is not necessary; all the `Format` method remains on the concrete type and the value is small.

kortschak commented on mat/format_value.go:
```diff
@@ -0,0 +1,143 @@
+	setSyntax(s int)
+}
+
+// vFormatOption is a functional option for value formatting.
```
Can we expand the `v` to `value` throughout please.

kortschak commented on mat/format_value.go:
```diff
@@ -0,0 +1,143 @@
+}
+
+// floatFormatter formats 64-bit a floating point value and satifies the
+// valueFormatter interface. A floatFormatter utilizes an internal buffer while
+// generating formatting output and this buffer persists with the formatter.
+// For this reason, and because a floatFormatter might be reused or revalued,
+// the floatFormatter embeds an exclusion lock.
```
Reuse doesn't require a lock, only if there is concurrent use, but concurrent use it not a goal, so the mutex can be dropped.

kortschak commented on mat/format_value.go:
```diff
@@ -0,0 +1,143 @@
+var _ valueFormatter = (*floatFormatter)(nil)
+
+// Format satisfies the fmt.Formatter interface.
+func (f *floatFormatter) Format(fs fmt.State, c rune) {
```
These methods don't need to be on a pointer.

kortschak commented on mat/format.go:
```diff
@@ -87,6 +87,52 @@ func (f formatter) Format(fs fmt.State, c rune) {
+// that satisfy the Matrix interface.
+type formattableMatrix struct{ mat Matrix }
+
+var _ formattable = (*formattableMatrix)(nil)
```
This should not check against the point since the methods are on the concrete type.

kortschak commented on mat/format.go:
```diff
@@ -87,6 +87,52 @@ func (f formatter) Format(fs fmt.State, c rune) {
+// tabular type is a two-dimensional array having rows and columns with table
+// elements accessible by a formattable. Details regarding value type and
+// methods for element access are left to the formattable.
+type tabular interface {
```
By convention we would probably call this `dimer` (or `dimser`) rather than `tabular`.

kortschak commented on mat/format.go:
```diff
@@ -87,6 +87,52 @@ func (f formatter) Format(fs fmt.State, c rune) {
+	Dims() (r, c int)
+}
+
+var _ tabular = (Matrix)(nil)
```
This is not necessary since `Matrix` is the canonical `interface{ Dim() (r, c int) }` type.

kortschak commented on mat/format.go:
```diff
@@ -427,58 +382,63 @@ func formatPython(m Matrix, prefix string, _ int, _ byte, squeeze bool, fs fmt.S
+// with the value in p, which may be updated for a new column by calling At.
+//
+// The fitted state is not safe for concurrent use.
+func (ws *state) fit(m formattable, printed int, squeeze bool, fs fmt.State, c rune, f fmt.Formatter) *state {
```
s/ws/s/g

kortschak commented on mat/format.go:
```diff
@@ -427,58 +382,63 @@ func formatPython(m Matrix, prefix string, _ int, _ byte, squeeze bool, fs fmt.S
-	)
+// fit fits a state to formattable m, storing a single width for the
+// formattable or, if squeeze is true, individual widths are stored for each
+// column in the formattable. After fitment, the state responds to Width calls
```
s/fitment/fitting/

kortschak commented on mat/format.go:
```diff
@@ -427,58 +382,63 @@ func formatPython(m Matrix, prefix string, _ int, _ byte, squeeze bool, fs fmt.S
+var _ fmt.State = (*state)(nil)
+
+// newState return a new state, shadowing given fmt.State fs.
+func newState(fs fmt.State) *state {
```
It's not clear to me why this needs to return a pointer type; the fit method returns the mutated `state`.  Either it can mutate in place and not bother returning the value (and having the remaining methods on the pointer), or preferably, returning the mutated non-pointer `state` (this would require that the `fmt.State` assignability check above changes).

kortschak commented on mat/format.go:
```diff
@@ -427,58 +382,63 @@ func formatPython(m Matrix, prefix string, _ int, _ byte, squeeze bool, fs fmt.S
+func (ws *state) fit(m formattable, printed int, squeeze bool, fs fmt.State, c rune, f fmt.Formatter) *state {
+	rows, cols := m.Dims()
+
+	// assign appropriate widther
```
Initial capital and final period for these comments.

kortschak commented on mat/format.go:
```diff
@@ -173,25 +173,25 @@ func format(m formattable, prefix string, margin int, dot byte, squeeze bool, fs
 		fmt.Fprintf(fs, "Dims(%d, %d)\n", rows, cols)
 	}
 
+	var buf = make([]byte, 0, 4)
```
`buf := make([]byte, 0, 4) // Four bytes required because of unicode decoration.`

kortschak commented on mat/format.go:
```diff
@@ -274,21 +274,21 @@ func formatMATLAB(m formattable, prefix string, _ int, _ byte, squeeze bool, fs
 		return
 	}
 
+	var buf = make([]byte, 0, 4)
```
I don't think this presizing to 4 makes sense here since in the default case `prefix` is arbitrarily long with a minumum of two. The remainder are one byte long, so maybe just presize to 2, `buf = make([]byte, 0, 2)` and let it grow if needed in the default case when `len(prefix) != 0`.

kortschak commented on mat/format.go:
```diff
@@ -361,24 +361,24 @@ func formatPython(m formattable, prefix string, _ int, _ byte, squeeze bool, fs
 		return
 	}
 
+	var buf = make([]byte, 0, 4)
```
`buf = make([]byte, 0, 3)`

kortschak commented on mat/format.go:
```diff
@@ -243,20 +243,20 @@ func formatMATLAB(m formattable, prefix string, _ int, _ byte, squeeze bool, fs
 			return
 		}
-		fs.Write([]byte{'['})
+		fmt.Fprint(fs, "[")
```
I'd prefer not to make these changes; they are more heavy weight. `fmt` depends on `reflect` and interpretation of types while write just spits bytes which the compiler has nicely prepared.

wamuir replied:
This was helpful, thanks.

kortschak commented on mat/format.go:
```diff
@@ -21,6 +21,19 @@ func Formatted(m Matrix, options ...FormatOption) fmt.Formatter {
+	for _, o := range options {
+		o(&f)
+	}
+	return &f
```
Return the non-pointer type.

kortschak commented on mat/format_value.go:
```diff
@@ -141,3 +141,113 @@ func formatFloat(v float64, buf []byte, dot byte, syntax int, fs fmt.State, c ru
+var _ valueFormatter = (*complexFormatter)(nil)
+
+// Format satisfies the fmt.Formatter interface.
+func (f *complexFormatter) Format(fs fmt.State, c rune) {
```
Here too, this doesn't need to be on a pointer for the same reasons.

kortschak commented on mat/format_value.go:
```diff
@@ -141,3 +141,113 @@ func formatFloat(v float64, buf []byte, dot byte, syntax int, fs fmt.State, c ru
+}
+
+// floatComplex formats 128-bit complex value and satifies the valueFormatter
+// interface. A complexFormatter utilizes an internal buffer while generating
+// formatting output and this buffer persists with the formatter. For this
+// reason, and because a complexFormatter might be reused or revalued,
+// the complexFormatter embeds an exclusion lock.
```
Same situation wrt the mutex.

wamuir commented:
> just a drive-by comment:
> could you run the benchmarks before/after the proposed optimization a couple of times (say 10-20) and show the result of `benchstat`?
> with the limited statistics you presented in the commit message, I get this rather poor statistical power result:
> 
> ```
> $> benchstat ./ref.txt ./new.txt 
> name        old time/op    new time/op    delta
> Format10      33.1µs ± 0%    54.6µs ± 0%   ~     (p=1.000 n=1+1)
> Format100     3.75ms ± 0%    7.07ms ± 0%   ~     (p=1.000 n=1+1)
> Format1000     334ms ± 0%     698ms ± 0%   ~     (p=1.000 n=1+1)
> 
> name        old alloc/op   new alloc/op   delta
> Format10      2.69kB ± 0%    2.89kB ± 0%   ~     (p=1.000 n=1+1)
> Format100     1.67MB ± 0%    1.67MB ± 0%   ~     (p=1.000 n=1+1)
> Format1000     164MB ± 0%     164MB ± 0%   ~     (p=1.000 n=1+1)
> 
> name        old allocs/op  new allocs/op  delta
> Format10        16.0 ± 0%      19.0 ± 0%   ~     (p=1.000 n=1+1)
> Format100        137 ± 0%        49 ± 0%   ~     (p=1.000 n=1+1)
> Format1000     1.07k ± 0%     0.08k ± 0%   ~     (p=1.000 n=1+1)
> ```
> 
> ie:
> 
> ```
> $> go test -run=NONE -bench=Format -benchmem -count=20 |& tee ref.txt
> ## apply optimizations
> $> go test -run=NONE -bench=Format -benchmem -count=20 |& tee new.txt
> $> benchstat ./ref.txt ./new.txt
> ```
> 
> thanks.
> 
> PS: `benchstat` is available here:
> 
>     * [golang.org/x/perf/benchstat](https://golang.org/x/perf/benchstat)

Thanks, I wasn't aware of benchstat.  Cool.  Output is below.  I added `-benchtime=1x`, because this appears to otherwise perform tests using the pooled mean?  If true, I'd think the dispersion of ± time/op is otherwise understated. I guess that's a nit on `go test` for not having a flag to output variance (or a nit on me for not being able to find it).

| name       | old time/op | new time/op  | delta                       |
| :--------- | ----------: | -----------: | --------------------------: |
| Format10   | 44.2µs ±17% |  73.9µs ±38% |  +67.07%  (p=0.000 n=20+18) |
| Format100  | 3.28ms ± 3% |  7.55ms ± 5% | +130.21%  (p=0.000 n=20+18) |
| Format1000 |  329ms ± 1% |   764ms ± 2% | +132.17%  (p=0.000 n=20+20) |

Increase in clock time is statistically significant.  @kortschak provided review comments, which I'll work through over the weekend and then rerun the bench.  One of these comments is to use micro-benchmarking and I expect this will offer greater insight into differences.  Addressing his comments may also close the current gap (e.g., his comment regarding fmt.Fprint and reflect).

wamuir commented:
Most recent commit addresses review comments, except for the switch to sub-benchmarks.  I'll need to push a second commit to update the benchmarks.  I attempted some further simplification of the format functions and I think this was productive.  I also added comments along the way.  Updated benchstat output is below.

| name       | old time/op | new time/op  | delta                       |
| :--------- | ----------: | -----------: | --------------------------: |
| Format10   | 44.9µs ±10% |  62.9µs ±31% |  +40.08%  (p=0.000 n=18+20) |
| Format100  | 3.31ms ± 1% |  4.27ms ± 1% |  +29.20%  (p=0.000 n=18+20) |
| Format1000 |  326ms ± 1% |   426ms ± 1% |  +30.41%  (p=0.000 n=20+20) |

wamuir commented:
| name                | old time/op | new time/op | delta                      |
| :------------------ | ----------: | ----------: | -------------------------: |
| Format/10/General   | 52.6µs ±17% | 68.8µs ±18% | +30.74%  (p=0.000 n=19+19) |
| Format/10/DotByte   | 49.5µs ±21% | 67.9µs ±17% | +37.19%  (p=0.000 n=19+20) |
| Format/10/Excerpt   | 50.3µs ±11% | 66.1µs ±21% | +31.40%  (p=0.000 n=19+20) |
| Format/10/Prefix    | 49.4µs ±10% | 66.4µs ±17% | +34.54%  (p=0.000 n=17+20) |
| Format/10/Squeeze   | 53.0µs ±22% | 69.7µs ±14% | +31.49%  (p=0.000 n=20+20) |
| Format/10/MATLAB    | 51.5µs ±12% | 66.4µs ±12% | +28.79%  (p=0.000 n=20+19) |
| Format/10/MATLAB#   | 43.0µs ±18% | 42.9µs ±23% |    ~     (p=0.967 n=20+20) |
| Format/10/Python    | 50.4µs ±13% | 64.0µs ± 9% | +26.99%  (p=0.000 n=20+19) |
| Format/10/Python#   | 42.4µs ±37% | 43.1µs ±30% |    ~     (p=0.757 n=20+20) |
| Format/100/General  | 3.13ms ± 7% | 4.43ms ± 9% | +41.42%  (p=0.000 n=17+20) |
| Format/100/DotByte  | 3.94ms ± 1% | 4.35ms ± 8% | +10.32%  (p=0.000 n=19+20) |
| Format/100/Excerpt  | 4.00ms ± 1% | 4.54ms ± 6% | +13.45%  (p=0.000 n=20+20) |
| Format/100/Prefix   | 4.00ms ± 1% | 5.61ms ± 1% | +40.23%  (p=0.000 n=20+16) |
| Format/100/Squeeze  | 4.42ms ± 1% | 5.35ms ±18% | +21.01%  (p=0.000 n=14+20) |
| Format/100/MATLAB   | 4.68ms ±15% | 6.46ms ±13% | +37.92%  (p=0.000 n=20+20) |
| Format/100/MATLAB#  | 2.91ms ± 3% | 2.90ms ± 2% |    ~     (p=0.247 n=20+20) |
| Format/100/Python   | 6.33ms ±16% | 6.38ms ±35% |    ~     (p=0.890 n=20+20) |
| Format/100/Python#  | 3.56ms ±13% | 1.69ms ± 1% | -52.45%  (p=0.000 n=20+19) |
| Format/1000/General |  306ms ± 0% |  405ms ± 0% | +32.08%  (p=0.000 n=19+20) |
| Format/1000/DotByte |  303ms ± 0% |  392ms ± 0% | +29.54%  (p=0.000 n=20+20) |
| Format/1000/Excerpt |  308ms ± 1% |  405ms ± 0% | +31.76%  (p=0.000 n=20+20) |
| Format/1000/Prefix  |  307ms ± 0% |  404ms ± 0% | +31.67%  (p=0.000 n=20+20) |
| Format/1000/Squeeze |  307ms ± 0% |  405ms ± 0% | +31.83%  (p=0.000 n=20+20) |
| Format/1000/MATLAB  |  307ms ± 0% |  405ms ± 0% | +31.99%  (p=0.000 n=19+20) |
| Format/1000/MATLAB# |  164ms ± 1% |  164ms ± 1% |    ~     (p=0.561 n=19+19) |
| Format/1000/Python  |  308ms ± 1% |  406ms ± 1% | +31.86%  (p=0.000 n=20+18) |
| Format/1000/Python# |  164ms ± 1% |  164ms ± 1% |    ~     (p=0.604 n=20+19) |
| [Geo mean]          |      3.74ms |      4.43ms | +18.29%                    |

wamuir commented:
| name                | old time/op | new time/op | delta                      |
| :------------------ | ----------: | ----------: | -------------------------: |
| Format/10/General   | 52.6µs ±17% | 68.8µs ±18% | +30.74%  (p=0.000 n=19+19) |
| Format/10/DotByte   | 49.5µs ±21% | 67.9µs ±17% | +37.19%  (p=0.000 n=19+20) |
| Format/10/Excerpt   | 50.3µs ±11% | 66.1µs ±21% | +31.40%  (p=0.000 n=19+20) |
| Format/10/Prefix    | 49.4µs ±10% | 66.4µs ±17% | +34.54%  (p=0.000 n=17+20) |
| Format/10/Squeeze   | 53.0µs ±22% | 69.7µs ±14% | +31.49%  (p=0.000 n=20+20) |
| Format/10/MATLAB    | 51.5µs ±12% | 66.4µs ±12% | +28.79%  (p=0.000 n=20+19) |
| Format/10/MATLAB#   | 43.0µs ±18% | 42.9µs ±23% |    ~     (p=0.967 n=20+20) |
| Format/10/Python    | 50.4µs ±13% | 64.0µs ± 9% | +26.99%  (p=0.000 n=20+19) |
| Format/10/Python#   | 42.4µs ±37% | 43.1µs ±30% |    ~     (p=0.757 n=20+20) |
| Format/100/General  | 3.13ms ± 7% | 4.43ms ± 9% | +41.42%  (p=0.000 n=17+20) |
| Format/100/DotByte  | 3.94ms ± 1% | 4.35ms ± 8% | +10.32%  (p=0.000 n=19+20) |
| Format/100/Excerpt  | 4.00ms ± 1% | 4.54ms ± 6% | +13.45%  (p=0.000 n=20+20) |
| Format/100/Prefix   | 4.00ms ± 1% | 5.61ms ± 1% | +40.23%  (p=0.000 n=20+16) |
| Format/100/Squeeze  | 4.42ms ± 1% | 5.35ms ±18% | +21.01%  (p=0.000 n=14+20) |
| Format/100/MATLAB   | 4.68ms ±15% | 6.46ms ±13% | +37.92%  (p=0.000 n=20+20) |
| Format/100/MATLAB#  | 2.91ms ± 3% | 2.90ms ± 2% |    ~     (p=0.247 n=20+20) |
| Format/100/Python   | 6.33ms ±16% | 6.38ms ±35% |    ~     (p=0.890 n=20+20) |
| Format/100/Python#  | 3.56ms ±13% | 1.69ms ± 1% | -52.45%  (p=0.000 n=20+19) |
| Format/1000/General |  306ms ± 0% |  405ms ± 0% | +32.08%  (p=0.000 n=19+20) |
| Format/1000/DotByte |  303ms ± 0% |  392ms ± 0% | +29.54%  (p=0.000 n=20+20) |
| Format/1000/Excerpt |  308ms ± 1% |  405ms ± 0% | +31.76%  (p=0.000 n=20+20) |
| Format/1000/Prefix  |  307ms ± 0% |  404ms ± 0% | +31.67%  (p=0.000 n=20+20) |
| Format/1000/Squeeze |  307ms ± 0% |  405ms ± 0% | +31.83%  (p=0.000 n=20+20) |
| Format/1000/MATLAB  |  307ms ± 0% |  405ms ± 0% | +31.99%  (p=0.000 n=19+20) |
| Format/1000/MATLAB# |  164ms ± 1% |  164ms ± 1% |    ~     (p=0.561 n=19+19) |
| Format/1000/Python  |  308ms ± 1% |  406ms ± 1% | +31.86%  (p=0.000 n=20+18) |
| Format/1000/Python# |  164ms ± 1% |  164ms ± 1% |    ~     (p=0.604 n=20+19) |
| [Geo mean]          |      3.74ms |      4.43ms | +18.29%                    |

vladimir-ch commented:
This PR has gone a bit stale. @wamuir are you still interested in submitting this change?

wamuir commented:
@vladimir-ch Yes. I think this needs another review. I can make revisions.
